### ISSUE TITLE
@ConfigurationProperties using Resource is resolved before S3 is available

### ISSUE THREAD

**Issue by osi** [spring-cloud/spring-cloud-aws#369]
Using `spring-cloud-aws-context 2.0.0.RELEASE`

I have a `@ConfigurationProperties` class that has an option of type `Resource`. In my configuration file, I have an S3 URL.

The resource was _not_ being returned as a `SimpleStorageResource`. I had to add `@DependsOn("org.springframework.cloud.aws.context.support.io.SimpleStorageProtocolResolverConfigurer#0")` to my configuration class to ensure that the S3 resolver was loaded first.

This was:
 - not obvious
 - hacky, as I'm depending on a generated bean name

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by osi** [comment]
I'll work on it, but can I have more than two weeks? 

It took two months before the bot added a label, and another 3 months to request a sample after that.

**Comment by spencergibb** [comment]
NP

**Comment by osi** [comment]
Here's a test. Remove the `@DependsOn` on `TestUserConfig` and it will not resolve the S3 resource.

Tested against the git tag for v2.1.0-RELEASE

(Thanks to @twicksell for helping start the example)

```
package org.springframework.cloud.aws.context.support.io;

import com.amazonaws.services.s3.AmazonS3;
import org.junit.Before;
import org.junit.Test;
import org.springframework.boot.autoconfigure.AutoConfigurations;
import org.springframework.boot.autoconfigure.context.ConfigurationPropertiesAutoConfiguration;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.boot.test.context.runner.ApplicationContextRunner;
import org.springframework.cloud.aws.context.config.annotation.ContextResourceLoaderConfiguration;
import org.springframework.cloud.aws.core.io.s3.SimpleStorageResource;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.DependsOn;
import org.springframework.core.io.Resource;

import static org.assertj.core.api.Assertions.assertThat;

public class ContextResourceLoaderConfigurationPropertiesTest {
    private ApplicationContextRunner contextRunner;

    @Before
    public void beforeTest() {
        contextRunner = new ApplicationContextRunner()
                            .withConfiguration(AutoConfigurations.of(ContextResourceLoaderConfiguration.class,
                                                                     ConfigurationPropertiesAutoConfiguration.class))
                            .withUserConfiguration(TestUserConfig.class)
                            .withPropertyValues("s3.file=s3://a-bucket/a-path")
        ;
    }

    @Test
    public void testS3ProtocolResolverPresent() {
        contextRunner.run(context -> {
            assertThat(context).hasNotFailed();
            assertThat(context).hasSingleBean(AmazonS3.class);
            assertThat(context).hasSingleBean(SimpleStorageProtocolResolverConfigurer.class);
            assertThat(context).hasSingleBean(S3ConfigProps.class);
            S3ConfigProps s3ConfigProps = context.getBean(S3ConfigProps.class);
            assertThat(s3ConfigProps.getFile())
                .isInstanceOf(SimpleStorageResource.class);
        });
    }
}

@ConfigurationProperties("s3")
class S3ConfigProps {

    private Resource file;

    public Resource getFile() {
        return file;
    }

    public void setFile(Resource file) {
        this.file = file;
    }
}

@Configuration
@EnableConfigurationProperties(S3ConfigProps.class)
@DependsOn("org.springframework.cloud.aws.context.support.io.SimpleStorageProtocolResolverConfigurer#0")
class TestUserConfig {


}
```

**Comment by maciejwalkowiak** [comment]
It has been fixed in 2.2.5
